pkg/route/api/validation/validation.go

```diff
@@ -111,8 +115,94 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
 	return allErrs
 }
 
+type blockVerifierFunc func(block *pem.Block) error
+
+func publicKeyBlockVerifier(block *pem.Block) error {
+	key, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return err
+	}
+	block.Type = "PUBLIC KEY"
+	block.Bytes, err = x509.MarshalPKIXPublicKey(key)
+	return err
+}
+
+func certificateBlockVerifier(block *pem.Block) error {
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return err
+	}
+	block.Type = "CERTIFICATE"
+	block.Bytes = cert.Raw
+	return err
+}
+
+func privateKeyBlockVerifier(block *pem.Block) error {
+	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err != nil {
+		key, err = x509.ParsePKCS1PrivateKey(block.Bytes)
+		if err != nil {
+			key, err = x509.ParseECPrivateKey(block.Bytes)
+			if err != nil {
+				return fmt.Errorf("block %s is not valid", block.Type)
+			}
+		}
+	}
+	switch t := key.(type) {
+	case *rsa.PrivateKey:
+		block.Type = "RSA PRIVATE KEY"
+		block.Bytes = x509.MarshalPKCS1PrivateKey(t)
+	case *ecdsa.PrivateKey:
+		block.Type = "ECDSA PRIVATE KEY"
+		block.Bytes, err = x509.MarshalECPrivateKey(t)
+	default:
+		err = fmt.Errorf("block private key %T is not valid", key)
+	}
+	return err
+}
+
+var knownBlockDecoders = map[string]blockVerifierFunc{
+	"RSA PRIVATE KEY":   privateKeyBlockVerifier,
+	"ECDSA PRIVATE KEY": privateKeyBlockVerifier,
+	"PRIVATE KEY":       privateKeyBlockVerifier,
+	"PUBLIC KEY":        publicKeyBlockVerifier,
+	// Potential "in the wild" PEM encoded blocks that can be normalized
+	"RSA PUBLIC KEY":   publicKeyBlockVerifier,
+	"DSA PUBLIC KEY":   publicKeyBlockVerifier,
+	"ECDSA PUBLIC KEY": publicKeyBlockVerifier,
+	"CERTIFICATE":      certificateBlockVerifier,
+}
+
+// sanitizePEM takes a block of data that should be encoded in PEM and returns only
+// the parts of it that parse and serialize as valid recognized certs in valid PEM blocks.
+// We perform this transformation to eliminate potentially incorrect / invalid PEM contents
+// to prevent OpenSSL or other non Golang tools from receiving unsanitized input.
+func sanitizePEM(data []byte) ([]byte, error) {
+	var block *pem.Block
+	buf := &bytes.Buffer{}
+	for len(data) > 0 {
+		block, data = pem.Decode(data)
+		if block == nil {
+			return buf.Bytes(), nil
+		}
+		fn, ok := knownBlockDecoders[block.Type]
```

Comment:
I'm wondering if we should just drop unrecognized block types and continue (as long as we got at least one recognized block)